landsat/anaconda-project.yml

```diff
@@ -10,38 +10,30 @@ examples_config:
 
 user_fields: [examples_config]
 
-channels: [pyviz, defaults]
+channels: [pyviz, defaults, conda-forge]
 
 packages: &pkgs
-- python=3.7
+- python
```

Comment:
I think our convention is to pin `python` in the project file

(There's a discussion somewhere on the GH repo where we even suggest that the convention could be to add a `>=` pin for all the packages - well except Python I guess - to track what minimum versions of each direct dependency were required to run the project vs. the dependencies that end up being locked up. As there's nothing to automate that process and it's kind of a pain to do manually since you then need to re-lock as anaconda-project finds out the project file changed, I doubt a little we'll manage to enforce that)

Reply:
I was going to ask if pinning each package in anaconda-project.yml is really need3ed since the anaconda-project-lock.yml already pins all packages; though that's not easy to read. I'll pin these though it would be nice if this could be automated somehow

Reply:
I've previously argued that you should put >= pins for any package whose API you depend on, filling in the version number of whatever you wrote it for. I.e. if you used pandas 1.2 when you wrote the notebook, pin to >=1.2 in anaconda-project.yml, since you may have used syntax only available in 1.2 or later. If you tested with pandas 1.1, then put >=1.1.  That way you have a recording of what you originally did, which can help explain things. E.g. if you're wondering why a particular notebook is using some weird old way of doing things, you can see that it was written for holoviews 1.2 and can realize "oh! they just didn't know any better then! I can go ahead and update that now". 

In practice, though, as @maximlt says above, most people don't do this, so I'm not sure if there's much point if it's not enforced and not done much in practice.

Reply:
**New learning:**
I understand. We don't specifically care about the `jupyter_client` dependency.
There should be some way to lock `notebook <x.x.x` and/or `python <x.x.x` such that it ensures the correct version of jupyter_client less than 8 is pulled. This is an interesting problem since we really want to know if there is a combination of package versions defined in `anaconda-project.yml` which pulls in jupyter_client <8.0. Currently this happens manually, trial and error AFAIK or using `conda search --info` but a quick test shows that it has a lower bound but not necessarily an upper bound. 🤔 

**Another followup from original question:**
I just wrote a simple script to get the min version out of the `anaconda-project-lock.yml` for any unpinned packages. This wasn't hard to do but unsure how robust it is - it just **finds the min version from all platforms and uses that as lower bound for each package that is not explicitly pinned**.  Then regenerate the lock file and test to ensure things still work on the 1 platform you're using. 

I think this should be obtained from the environment or from the lock file since presumably the env is a reflection of the lock file. Nicer if it is scripted, plus less error prone. But is the above claim in **bold** always correct? We can certainly (re)-test. Tools should make our life easier, so definitely not advocating to creating tools that become complex and must then be maintained.

The requirement of adding the bound in the `anaconda-project.yml` makes it easy to see min version of packages without going to the lock file which is nice for sure. I think we can automate this assuming that retesting after generating the lower bound results in the notebook working as before (99% of the time).

I can move these out as separate issues/topics to discuss since they apply more generally to all examples.

Reply:
Resolving for now.